### PR TITLE
add un_balance to multi class

### PR DESCRIPTION
**Pull request by guolinke**
refer to #306

### CODE REVIEW

**Comment by guolinke** [comment]
@cbecker

File: src/objective/multiclass_objective.hpp [hunk 1]
```diff
@@ -24,12 +25,32 @@ class MulticlassLogloss: public ObjectiveFunction {
...
+    label_neg_weights_ = std::vector<float>(num_class_, 1);
+    if (is_unbalance_) {
+      std::vector<int> cnts(num_class_, 0);
+      for (int i = 0; i < num_data_; ++i) {
+        ++cnts[label_int_[i]];
+      }
+      int average_cnt = num_data_ / num_class_;
+      for (int i = 0; i < num_class_; ++i) {
+        int cnt_cur = cnts[i];
+        int cnt_other = (average_cnt - cnts[i]);
```
Comment (wxchan) [inline src/objective/multiclass_objective.hpp]: should `average_cnt` be `num_data_` here? `cnt_other`, as well as `weights`, can be negative for current algorithm.

**Comment by cbecker** [comment]
I am not very sure that modifying the weights in the binary classifiers (if I am reading this right) is the formal way of doing this. I would rather compute the weights of each instance as `w[i] = Ntotal / (N[j] * numClasses)` where 

```
Ntotal = total number of samples (all classes)
N[j] = number of samples for class j
numClasses = number of classes
```

This comes from forcing the weight of a sample to be inversely proportional to the number of samples in its class, plus that the sum of the weights equals `Ntotal`.

**Comment by guolinke** [comment]
@wxchan @cbecker 
it makes sense. I will update it.

File: src/objective/multiclass_objective.hpp [hunk 2]
```diff
@@ -46,11 +60,12 @@ class MulticlassLogloss: public ObjectiveFunction {
           auto p = rec[k];
           size_t idx = static_cast<size_t>(num_data_) * k + i;
           if (label_int_[i] == k) {
-            gradients[idx] = static_cast<score_t>(p - 1.0f);
+            gradients[idx] = static_cast<score_t>(p - 1.0f) * label_pos_weights_[k];
+            hessians[idx] = static_cast<score_t>(2.0f * p * (1.0f - p))* label_pos_weights_[k];
           } else {
             gradients[idx] = static_cast<score_t>(p);
+            hessians[idx] = static_cast<score_t>(2.0f * p * (1.0f - p));
```
Comment (cbecker) [inline src/objective/multiclass_objective.hpp]: I would expect gradients and hessians to be multiplied by the weight corresponding to their class. Which is the reason behind only doing so for the positive samples in this binary classifier?

I would like to dig deeper into the loss itself. Is this the gradient from the loss function at https://github.com/Microsoft/LightGBM/blob/b884da85f1234e12b77f09ce054b6c43b0329454/src/metric/multiclass_metric.hpp ?

Comment (guolinke) [inline src/objective/multiclass_objective.hpp]: @cbecker 
I think your idea is also a solution. But i am not sure which one will be better. 
My idea is to multiply a big weight to gradient/hessian with small-frequency class, so that the model will give bigger prediction to them.
Your idea is to give sample weights for different class, and softmax will give bigger prediction for these larger-weight data.

Comment (cbecker) [inline src/objective/multiclass_objective.hpp]: We can try different heuristics. I was thinking of using the loss itself as the departure point, and from there derive what the hessian and gradient should be like. I would be happy to try both.

Now I am a bit busy with other things, but the loss-based one I discussed about is quick to implement. I will try to make some time today or early next week.

**Comment by guolinke** [comment]
@cbecker I merge this first. you can create a new PR after tested.

**Comment by cbecker** [comment]
@guolinke ok thanks. Did you try on some datasets? How is it doing?

**Comment by akaniklaus** [comment]
@guolinke How I can try this? I have a multi-class classification problem, ligthGBM is just predicting nothing but the most frequent class.

**Comment by khrisanfov** [comment]
@akaniklaus Hi, I am not sure but I think you should set is_unbalance to True and objective to multiclassova (not multiclass), because is_unbalance works only with binary classification. If you found better solution please tell me, I am interested in it too.

**Comment by akaniklaus** [comment]
@khrisanfov Yes, I also did that way at the end and now just noticed that one may need the same functionality for regression as well.
